ballerina/variable_validator.bal

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import graphql.parser;
+
+class VariableValidator{
+    *parser:Visitor;
+
+    private parser:DocumentNode documentNode;
+    private map<json> variables;
+    private string[] visitedVariableDefinitions;
+    private ErrorDetail[] errors;
+
+    public isolated function init(parser:DocumentNode documentNode, map<json>? variableValues){
+        self.documentNode = documentNode;
+        self.visitedVariableDefinitions = [];
+        self.errors = [];
+        if variableValues == () {
+            self.variables = {};
+        } else {
+            self.variables = <map<json>> variableValues;
+        }
+    }
+
+    public isolated function validate() returns ErrorDetail[]? {
+        self.visitDocument(self.documentNode);
+        if self.errors.length() > 0 {
+            return self.errors;
+        }
+    }
+
+    public isolated function visitDocument(parser:DocumentNode documentNode, anydata data = ()) {
+        parser:OperationNode[] operations = documentNode.getOperations();
+        foreach parser:OperationNode operationNode in operations {
+            foreach ErrorDetail errorDetail in operationNode.getErrors() {
+                self.errors.push(errorDetail);
+            }
+            self.visitOperation(operationNode);
+        }
+    }
+
+    public isolated function visitOperation(parser:OperationNode operationNode, anydata data = ()) {
+        map<parser:VariableDefinition> variableDefinitions = operationNode.getVaribleDefinitions();
+        foreach parser:Selection selection in operationNode.getSelections() {
+            self.visitSelection(selection, variableDefinitions);
+        }
+        string[] keys = variableDefinitions.keys();
+        foreach string name in keys {
+            if self.visitedVariableDefinitions.indexOf(name) == () {
```

Comment:
Seems like this introduces a bug. When we keep the `visitedVariableDefinitions` in global scope, we have a single array of `visitedVariableDefinitions` for the document. But we define variables in the scope of operation.

In this case, when an operation contains an unused variable, but that same variable is used in another operation, this will not throw an error. Is that the expected behavior? Otherwise, we have to update this.

Consider the following example:

```
query getCity($id: Int, $extra: String) {
    profileExtra(id: $id, extra: $extra) {
        city
    }
}

query getNameById($id: Int, $name: String) {
    profile(id: $id) {
        name
    }
}
```

This should return an error since `getNameById` operation has an unused variable. This might depend on the operation order too.

Reply:
Let's add a test case for this scenario